### PR TITLE
Weighted objective function

### PR DESCRIPTION
**Pull request by brynpickering**
Fixes issue(s) #187

Summary of changes in this pull request:

* `run.objective_options.cost_class` can now be defined as a dictionary of cost_class:weight key:value pairs
* objective will consider all cost classes in `run.objective_options.cost_class` and will weight them according to the weights given. If still given as a string or given as a set, weight(s) will be considered as 1

Reviewer checklist:

- [x] Test(s) added to cover contribution
- [x] Documentation updated
- [x] Changelog updated
- [x] Coverage maintained or improved

### CODE REVIEW

File: calliope/config/model.yaml [hunk 1]
```diff
@@ -14,7 +14,7 @@ run:
     cyclic_storage: true # If true, storage in the last timestep of the timeseries is considered to be the 'previous timestep' in the first timestep of the timeseries
     ensure_feasibility: false # If true, unmet_demand will be a decision variable, to account for an ability to meet demand with the available supply. If False and a mismatch occurs, the optimisation will fail due to infeasibility
     mode: plan  # Which mode to run the model in: 'plan' or 'operation'
-    objective_options: {'cost_class': 'monetary', 'sense': 'minimize'}  # Arguments to pass to objective function. If cost-based objective function in use, should include 'cost_class' and 'sense' (maximize/minimize)
+    objective_options: {'cost_class': {'monetary'}, 'sense': 'minimize'}  # Arguments to pass to objective function. If cost-based objective function in use, should include 'cost_class' and 'sense' (maximize/minimize)
```
Comment (sjpfenninger) [inline calliope/config/model.yaml]: I think this should be `{'cost_class': {'monetary': 1}, 'sense': 'minimize'}`

Comment (brynpickering) [reply calliope/config/model.yaml]: I'm not too fussed, they all end up as the same thing in the end.

**Comment by brynpickering** [comment]
Will add tests in next few hours

**Comment by brynpickering** [comment]
So, there are a few issues outstanding with this feature:

1. `calliope.exceptions.ModelWarning` is expected, based on a check for cost classes defined as sets, which are then converted to dicts, with the key as the cost class and the value equal to one. Current tests fail because this isn't being caught, or the warning being raised isn't as expected.

2. The way `AttrDict.union()` combines two dictionaries doesn't work with this feature. Either:
a. the default is `objective_options.cost_class: {'monetary' : 1}` which is not overridden on the user saying, e.g. `objective_options.cost_class: {'emissions' : 1}`, instead you get `objective_options.cost_class: {'monetary' : 1, 'emissions' : 1}`. One would have to explicitly include `'monetary': 0` to remove it.
b. the default is `objective_options.cost_class: 'monetary'` which, on overriding with e.g. `objective_options.cost_class: {'emissions' : 1}`, completely removes other keys in the `objective_options` dict, creating `objective_options: {'cost_class': {'emissions' : 1}}`. I.e. it removes `sense`. This causes the model to fail, since it expects `sense` to be there.

**Comment by brynpickering** [comment]
Note @sjpfenninger and @timtroendle : I've updated all `DeprecationWarning` to `FutureWarning`. The former is ignored by default in Python interpreters, so users would not get these messages. We should keep `DeprecationWarning` for developer facing functions which are deprecated. `FutureWarning` could instead be `calliope.exceptions.ModelWarning`, if you prefer, but then it might be difficult to extract those warnings for ignoring in pytest...

**Pull request merged**